Update crypto fork to v0.6.0-1 tag

jentfoo opened the pull request:
This tag includes the most recent updates from the upstream in order to address recently fixed CVE's.

Changes included: https://github.com/gravitational/crypto/compare/903e65687e59...c3f983bc73dcf711908e49d91bda9cffbecb13cd

jentfoo commented on go.sum:
```diff
@@ -1418,7 +1418,6 @@ golang.org/x/net v0.0.0-20220624214902-1bab6f366d9e/go.mod h1:XRhObCWvk6IyKnWLug
 golang.org/x/net v0.0.0-20220630215102-69896b714898/go.mod h1:XRhObCWvk6IyKnWLug+ECip1KBveYUHfp+8e9klMJ9c=
 golang.org/x/net v0.0.0-20220722155237-a158d28d115b/go.mod h1:XRhObCWvk6IyKnWLug+ECip1KBveYUHfp+8e9klMJ9c=
 golang.org/x/net v0.1.0/go.mod h1:Cx3nUiGt4eDBEyega/BKRp+/AlGL8hYe7U9odMt2Cco=
-golang.org/x/net v0.3.0/go.mod h1:MBQ8lrhLObU/6UmLb4fmbmk5OcyYmqtbGd/9yIeKjEE=
```
Removed after executing `go mod tidy`

jakule reviewed (changes requested):
I don't mind the changes, but SSH access needs to be tested before merging it.

jentfoo commented:
It turns out the flagged dependency alerts were a false positive due to our fork.  So there is no urgency with this update, but I still believe it would be good to get it across the line since I started it.

A chance in crypto started to return errors if input was over 72 characters (rather than ignoring the extra input).  I will make an update to maintain compatibility with our prior behavior.

jakule commented on lib/utils/bcrypt.go:
```diff
@@ -0,0 +1,37 @@
+	return input
+}
+
+func BcryptFromPassword(password []byte, cost int) ([]byte, error) {
```
Missing Go Doc. This is a public method, and it should have documentation.

jentfoo replied:
Documentation added

jakule commented on lib/utils/bcrypt.go:
```diff
@@ -0,0 +1,37 @@
+// from ignoring the extra input to returning an error, this truncation is necessary to maintain compatibility with
+// customers who have long passwords, or more commonly our recovery codes.
+func truncateToMaxSize(input []byte) []byte {
+	if len(input) > maxInputSize {
+		return input[:maxInputSize]
+	}
```
I don't think this is the correct way of dealing with this issue. I agree that we should truncate passwords that are already in our DB, but I also think we should return the error when someone tries to set a password that is longer than 72 characters. Otherwise, we give people a false sense of security.
Another question is, do we really want to limit the password to 72 chars? In my opinion, sounds reasonable, tbh

I also think that we should mention in our documentation what passwords we support. 
WDYT?

jentfoo replied:
> I agree that we should truncate passwords that are already in our DB

At first I expected to need to do this for verifying passwords also, however the cryto implementation does still silently truncate when verifying.  So I don't believe any changes are necessary for supporting existing existing passwords and recovery tokens.

> I also think we should return the error when someone tries to set a password that is longer than 72 characters

I was on the fence about this.  You could argue that by continuing to accept truncated passwords we are perpetuating a false sense of security either way (but this is necessary to support existing passwords and reset tokens).  I felt like this implementation is the most consistent, but I agree with the reasons crypto made the changes to start returning an error on larger inputs (I just wish they had started that way).

That said, it's also worth recognizing that 72 characters is more than enough entropy to prevent any realistic brute force attack.  Although more characters is theoretically better, there is certainly an argument that it's also unnecessary.

If there is a preference to instead prevent passwords and recovery tokens > 72 characters I am happy to make that change.

jakule replied:
@r0mant @zmb3 Any thoughts? Do we want to:
1. Block people from using passwords longer than 72 chars by returning an error if someone tries to do it?
2. Keep truncating as we used to.

jakule reviewed (changes requested):
I tested the SSH access, and everything seems to work fine. I'm leaving "change request" to re-think the password handling.

jentfoo commented:
PR was rebased on master, @jakule can you re-review this?  Thank you!

jentfoo commented:
@r0mant and @zmb3 would you mind reviewing this change?  There is no security fix here, just generally looking to update this dependent fork so we maintain current and future updates remain straight forward.

jakule commented:
@r0mant @zmb3 Friendly ping.

jentfoo committed:
Update crypto fork to v0.6.0-1 tag

This tag includes the most recent updates from the upstream in order to address recently fixed CVE's.
Changes included: https://github.com/gravitational/crypto/compare/903e65687e59...c3f983bc73dcf711908e49d91bda9cffbecb13cd

In addition this adds `utils/bcrypto.go` to restore the prior crypto behavior of truncating input > 72 chars.
This commit in Crypto breaks functionality with long passwords or more commonly our reset tokens which are > 72 characters: https://github.com/gravitational/crypto/commit/bc7d1d1eb54b3530da4f5ec31625c95d7df40231
As described in the commit message, the prior behavior was to ignore the extra data.  Since 72 bytes are plenty of data to prevent brute force attacks, there is not a significant need to report this error to the user.

espadolini commented:
I'm getting failures in `make grpc` due to `api/go.mod` having a different replacement version for `golang.org/x/crypto`.

espadolini commented on lib/utils/bcrypt.go:
```diff
@@ -0,0 +1,39 @@
+// from ignoring the extra input to returning an error, this truncation is necessary to maintain compatibility with
+// customers who have long passwords, or more commonly our recovery codes.
+func truncateToMaxSize(input []byte) []byte {
+	if len(input) > maxInputSize {
+		return input[:maxInputSize]
+	}
```
Something like `input[:subtle.ConstantTimeSelect(subtle.ConstantTimeLessOrEq(len(input), 72), len(input), 72)]`

jentfoo replied:
This PR is not attempting to pretend that additional input is calculated as part of the hash.  So using a timing attack to determine that the password is truncated does not provide any value (the truncation is not supposed to be a secret process).

Let me know if you see another risk I may not have considered.  Thank you @espadolini!

espadolini replied:
I suppose the only thing you'd learn is that the password was intended to be longer than 72 and is now 72, versus the password being between 1 and 71 characters, which doesn't really help.

jentfoo replied:
And as an attacker that information does not help you, since you would only need to target the first 72 characters anyways

jentfoo replied:
Well, correction, it would help in knowing that you would need to target exactly 72 characters.  Although brute force of a password of that length is infeasible.

codingllama commented:
> I'm getting failures in `make grpc` due to `api/go.mod` having a different replacement version for `golang.org/x/crypto`.

+1, I ended up fixing it in https://github.com/gravitational/teleport/pull/23133. Curious how it managed to clear CI here.
